### PR TITLE
THRIFT-4797: Golang: Fix import collisions

### PR DESCRIPTION
**Pull request by johnboiles**
See [THRIFT-4797](https://issues.apache.org/jira/browse/THRIFT-4797) for details. In short, Golang compilation fails when a Thrift file includes two files that have the same final namespace component (e.g. `namespace go packagea.common` and `namespace go packageb.common`)

Somewhat based on [fbthrift's implementation](https://github.com/facebook/fbthrift/commit/b6754fe7a6a6e00ce288f14cd4e88cfa3d0cc3e4#diff-803353e952116d4e88caab467a62f35b) though their implementation simply always adds an alias with a unique numerical prefix.

- [x] Did you create an [Apache Jira](https://issues.apache.org/jira/projects/THRIFT/issues/) ticket?  (not required for trivial changes)
- [x] Does your pull request title follow the pattern "THRIFT-NNNN: describe my issue"?  (not required for trivial changes)
- [x] Did you squash your changes to a single commit?
- [x] Did you do your best to avoid breaking changes?  If one was needed, did you label the Jira ticket with "Breaking-Change"

### CODE REVIEW

File: compiler/cpp/src/thrift/generate/t_go_generator.cc [hunk 1]
```diff
@@ -809,17 +812,35 @@ string t_go_generator::render_includes(bool consts) {
...
+        // packageacommon as the alias for packagea/common instead of common=). But just
+        // appending an integer is much simpler code
+        value = tmp(value);
+      }
+      package_identifiers_set_.insert(value);
+      it = package_identifiers_.emplace(go_module, std::move(value)).first;
+    }
+    auto const& package_identifier = it->second;
+
+    result += "\t" + package_identifier + " \"" + gen_package_prefix_ +
```
Comment (johnboiles) [inline compiler/cpp/src/thrift/generate/t_go_generator.cc]: Actually it'd be better if I didn't include the alias if it's not necessary. I'll go change that.

**Comment by johnboiles** [comment]
Just found another case to handle here. I should also detect collisions with the golang packages that get [imported](https://github.com/apache/thrift/blob/master/compiler/cpp/src/thrift/generate/t_go_generator.cc#L875) like `bytes` `context` `reflect`

**Comment by dcelasun** [comment]
For stdlib collisions like `context` etc, it would be great if the import list was a vector instead of a hardcoded string so we could reuse it everywhere.

**Comment by johnboiles** [comment]
@dcelasun I made it use a vector for system imports can you take a look? I’ll squash it together if/when it looks good to you!

**Comment by johnboiles** [comment]
@dcelasun squashed and ready to merge!

**Comment by johnboiles** [comment]
Actually hold on merging for a bit. I _may_ have found another edge case

**Comment by johnboiles** [comment]
Yep, confirmed. If the package has a service, and that service’s last namespace component conflicts with a previously imported package, it won’t get aliased. Hopefully I can get a chance to fix tomorrow

**Comment by johnboiles** [comment]
@dcelasun one more commit to review to fix the issue with the remote package import collisions! Hopefully this is it

**Comment by dcelasun** [comment]
LGTM. Tried testing a few different scenarios locally, everything works fine. Thanks for the PR!

**Pull request merged**